admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -0,0 +1,188 @@
+            network_sender,
+            peer_info: upstream_peers,
+            client_events,
+            is_validator,
+        }
+    }
+
+    /// main routine. starts sync coordinator that listens for CoordinatorMsg
+    pub async fn process_network_messages(
+        mut self,
+        network_events: Vec<AdmissionControlNetworkEvents>,
+    ) {
+        let mut events = select_all(network_events).fuse();
+
+        loop {
+            ::futures::select! {
+                mut msg = self.client_events.select_next_some() => {
+                    if msg.has_submit_transaction_request() {
+                        if let Err(err) = self.submit_transaction_upstream(msg.take_submit_transaction_request()).await {
+                            error!("[upstream proxy] failed to send transaction upstream, error: {:?}", err);
+                        }
+                    };
+                },
+                network_event = events.select_next_some() => {
+                    match network_event {
+                        Ok(event) => {
+                            match event {
+                                Event::NewPeer(peer_id) => {
+                                    debug!("[ac upstream proxy] new peer {}", peer_id);
+                                    self.new_peer(peer_id);
+                                }
+                                Event::LostPeer(peer_id) => {
+                                    debug!("[ac upstream proxy] lost peer {}", peer_id);
+                                    self.lost_peer(peer_id);
+                                }
+                                Event::RpcRequest((peer_id, mut message, _)) => {
+                                    if message.has_submit_transaction_request() {
+                                        if let Err(err) = self.process_submit_transaction_request(message.take_submit_transaction_request()).await {
+                                            error!("[upstream proxy] failed to process transaction request, peer: {}, error: {:?}", peer_id, err);
+                                        }
+                                    }
+                                    if message.has_submit_transaction_response() {
+                                       self.process_submit_transaction_response(&peer_id, message.take_submit_transaction_response()).await;
+                                    }
+                                }
+                                _ => {},
+                            }
+                        },
+                        Err(err) => { error!("[ac upstream proxy] network error {:?}", err); },
+                    }
+                },
+            }
+        }
+    }
+
+    /// new peer discovery handler
+    /// adds new entry to `peer_info`
+    fn new_peer(&mut self, peer_id: PeerId) {
+        self.peer_info
+            .entry(peer_id)
+            .or_insert(PeerSyncState { is_alive: true })
```

Comment:
I don't think you want to insert new entry. If it's not upstream in config, you don't want to add it here